From a0fe192165e5454c65499cb988ef8cb39dd165e3 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Mon, 15 Apr 2019 15:22:52 +0800
Subject: [PATCH 21/30] Revert "MLK-18113: rpmsg: virtio_rpmsg_bus: fix
 rpmsg_probe() for virtio-mmio transport"

This reverts commit 51d5c7a64236a7b35d0c58f3550310a2d83bb9cd.

When booting kernel with DEBUG_PREEMPT enabled, it dumps the
warning as below:
WARNING: CPU: 1 PID: 1 at ./include/linux/dma-mapping.h:516 rpmsg_probe+0x400/0x4f0
Modules linked in:
CPU: 1 PID: 1 Comm: swapper/0 Not tainted 4.18.31-yocto-standard+ #450
Hardware name: Freescale i.MX8MQ EVK (DT)
pstate: 20000005 (nzCv daif -PAN -UAO)
pc : rpmsg_probe+0x400/0x4f0
lr : rpmsg_probe+0x12c/0x4f0
sp : ffff00000803b7b0
x29: ffff00000803b7b0 x28: 0000000000000000
x27: 0000000000040000 x26: 0000000000040000
x25: ffff800078846010 x24: ffff800078930e40
x23: 0000000000000000 x22: ffff000008fb8000
x21: ffff0000090c78c8 x20: ffff000008bb7000
x19: ffff800078930e00 x18: ffffffffffffffff
x17: 0000000000000000 x16: 0000000000000000
x15: ffff000008fb8588 x14: ffff000008e23000
x13: ffff8000786c01a0 x12: 0000000000000000
x11: ffff8000bfffe008 x10: 0000fffff6f17840
x9 : 0000000000000000 x8 : ffff80007894b090
x7 : 0000000000000000 x6 : 000000000000003f
x5 : 0000000000000040 x4 : fffffffffffffff0
x3 : 0000000000000001 x2 : 0000000000000000
x1 : 0000000000000200 x0 : 0000000000000000
Call trace:
 rpmsg_probe+0x400/0x4f0
 virtio_dev_probe+0x168/0x210
 driver_probe_device+0x224/0x300
 __device_attach_driver+0xa0/0xe8
 bus_for_each_drv+0x74/0xd8
 __device_attach+0xe0/0x138
 device_initial_probe+0x24/0x30
 bus_probe_device+0x9c/0xa8
 device_add+0x3d4/0x5f0
 register_virtio_device+0xf0/0x108
 imx_rpmsg_probe+0x430/0x5b8
 platform_drv_probe+0x58/0xa8
 driver_probe_device+0x224/0x300
 __driver_attach+0xe4/0xe8
 bus_add_driver+0x1c4/0x230
 driver_register+0x64/0x110
 __platform_driver_register+0x54/0x60
 imx_rpmsg_init+0x24/0x58
 do_one_initcall+0x68/0x258
 kernel_init_freeable+0x2e4/0x3e8
 kernel_init+0x18/0x110
 ret_from_fork+0x10/0x1c
---[ end trace 05a5e7f1f8d84511 ]---

Revert this commit because the upstream commit d999b622fcfb
("rpmsg: virtio: allocate buffer from parent") is a better solution.

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/rpmsg/virtio_rpmsg_bus.c | 18 +++++-------------
 1 file changed, 5 insertions(+), 13 deletions(-)

diff --git a/drivers/rpmsg/virtio_rpmsg_bus.c b/drivers/rpmsg/virtio_rpmsg_bus.c
index 2b7fd2a445fe..664f957012cd 100644
--- a/drivers/rpmsg/virtio_rpmsg_bus.c
+++ b/drivers/rpmsg/virtio_rpmsg_bus.c
@@ -56,7 +56,6 @@
 struct virtproc_info {
 	struct virtio_device *vdev;
 	struct virtqueue *rvq, *svq;
-	struct device *bufs_dev;
 	void *rbufs, *sbufs;
 	unsigned int num_bufs;
 	unsigned int buf_size;
@@ -917,16 +916,9 @@ static int rpmsg_probe(struct virtio_device *vdev)
 				     total_buf_space, &vrp->bufs_dma,
 				     GFP_KERNEL);
 	if (!bufs_va) {
-		bufs_va = dma_alloc_coherent(vdev->dev.parent,
-					     total_buf_space, &vrp->bufs_dma,
-					     GFP_KERNEL);
-		if (!bufs_va) {
-			err = -ENOMEM;
-			goto vqs_del;
-		} else
-			vrp->bufs_dev = vdev->dev.parent;
-	} else
-		vrp->bufs_dev = vdev->dev.parent->parent;
+		err = -ENOMEM;
+		goto vqs_del;
+	}
 
 	dev_dbg(&vdev->dev, "buffers: va %p, dma %pad\n",
 		bufs_va, &vrp->bufs_dma);
@@ -988,7 +980,7 @@ static int rpmsg_probe(struct virtio_device *vdev)
 	return 0;
 
 free_coherent:
-	dma_free_coherent(vrp->bufs_dev, total_buf_space,
+	dma_free_coherent(vdev->dev.parent->parent, total_buf_space,
 			  bufs_va, vrp->bufs_dma);
 vqs_del:
 	vdev->config->del_vqs(vrp->vdev);
@@ -1023,7 +1015,7 @@ static void rpmsg_remove(struct virtio_device *vdev)
 
 	vdev->config->del_vqs(vrp->vdev);
 
-	dma_free_coherent(vrp->bufs_dev, total_buf_space,
+	dma_free_coherent(vdev->dev.parent->parent, total_buf_space,
 			  vrp->rbufs, vrp->bufs_dma);
 
 	kfree(vrp);
-- 
2.17.1

